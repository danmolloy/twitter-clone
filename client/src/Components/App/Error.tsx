import { GlobeIcon, RefreshIcon } from "@heroicons/react/outline"

export const Error = () => {

  const handleClick = () => {
    alert(Object.keys(localStorage))
    window.location.reload()
  }

  return (
    <div id="error-page" className="w-full flex flex-col items-center mt-12">
      <GlobeIcon className="h-6 w-auto"/>
      <p className="text-gray-600">An error occurred. Please check your connection and try again.</p>
      <button className="tweet-btn py-1 flex flex-row" onClick={() => handleClick()}>
        <RefreshIcon className="h-6 w-auto"/>
        <p className="font-semibold px-1">Retry</p>
      </button>
    </div>
  )
}