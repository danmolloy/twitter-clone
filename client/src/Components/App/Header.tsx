export const Header = (props: any) => {
  return (
    <div className="w-full p-2 border-b border-r">
      <h2 className="text-xl font-semibold">{props.pageTitle}</h2>
      {props.blurb &&
      <p>{props.blurb}</p>}
    </div>
  )
}