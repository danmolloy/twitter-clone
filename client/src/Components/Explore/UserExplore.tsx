import { Link } from "react-router-dom"
import { ExploreUser } from "../../types"
import { FollowButton } from "../Profile/FollowButton"

export const UserExplore = (props: {user: ExploreUser, currentUserHandle: string | undefined}) => {
  return (
       <div className="border-b flex flex-row justify-between hover:bg-gray-100 ">
        <Link to={`/${props.user.handle.slice(1)}`} className="flex flex-row m-4">
        <img src={props.user.profilePic} className="w-12 sm:w-16 h-auto rounded-full " />
        <div className="flex flex-col ml-4">
          <h3 className="font-semibold">{props.user.name}</h3>
          <p className="text-gray-500">{props.user.handle}</p>
          <p>{props.user.blurb}</p>
        </div>
        </Link>
        <FollowButton currentUserHandle={props.currentUserHandle && props.currentUserHandle} user={props.user}/>
      </div>)
}