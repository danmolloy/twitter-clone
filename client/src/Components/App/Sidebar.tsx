import { HashtagIcon, 
  HomeIcon, 
  UserIcon, 
  BellIcon,
  InboxIcon,
  LogoutIcon,
} from '@heroicons/react/outline'
import { Link } from 'react-router-dom'
import { AUTH_TOKEN } from '../../constants'
import { gql, useQuery } from '@apollo/client'
import { Error } from './Error'
import { Loading } from './Loading'
import { useState } from 'react'
import { UserOptions } from './UserOptions'
import { Message, Notification } from '../../types'

export const DELETE_USER = gql`
  mutation SignUpMutation {
    deleteUser {
      name
    }
  }
`;

export const GET_USER = gql`
  query {
    getNotifications {
      notifications {
          read
        }
        chats {
          content {
            read
            authorHandle
          }
        }
        handle
        profilePic
    }
  }
`;


export const Sidebar = () => {
  const [showUserOptions, setShowUserOptions] = useState(false)

  const { loading, error, data } = useQuery(GET_USER)

  const onSignOut = () => {
    localStorage.removeItem(AUTH_TOKEN)
    window.location.reload()
  }

  if (error) {
    console.log(error)
    return <Error />
  }

  if (loading) {
    return <Loading />
  }

  return (
    <div id="side-bar" 
    className="side-bar">
      <Link to="/home" className="md:ml-2" id="twitter-home-link" title="Home">
        <img src="/favico.ico" alt="Blue Twitter bird" className="hidden sm:flex w-12 h-auto mx-6 my-4 p-2 rounded hover:bg-blue-50"/>
      </Link>
      <Link to="/home" className="lg-side-icon" id="home-link" title="Home">
        <HomeIcon className="side-icon"/>
        <label htmlFor="home-link" className="sidebar-text">Home</label>
      </Link>
      <Link to="/explore" title="Explore" className="lg-side-icon" id="explore-link">
        <HashtagIcon className="side-icon"/>
        <label htmlFor="explore-link" className="sidebar-text">Explore</label>
      </Link>
      <Link to="/notifications" title="Notifications" className="lg-side-icon" id="notifications-link">
      {data && data.getNotifications.notifications.map(
        (i: Notification) => i.read).includes(false) 
      && <div id="notifications-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <BellIcon className="side-icon"/>
        <label htmlFor="notifications-link" className="sidebar-text">Notifications</label>
      </Link>
      <Link to="/messages" title="Messages" className="lg-side-icon">
      {data && data.getNotifications.chats[0] !== undefined
      && data.getNotifications.chats[0].content.filter(
        (i:Message) => i.read === false && i.authorHandle !== data.getNotifications.handle).length > 0
        && <div id="unread-msg-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <InboxIcon className="side-icon"/>
        <label htmlFor="messages-link" className="sidebar-text">Messages</label>
      </Link>
      <button className="sm:hidden lg-side-icon" title="User options" onClick={() => setShowUserOptions(!showUserOptions)} >
        {showUserOptions 
        && <UserOptions 
        currentUser={data && data.getNotifications} 
        onSignOut={() => onSignOut()} 
        close={() => setShowUserOptions(false)}/>}
        <img alt={data.getNotifications?.name ? data.getNotifications?.name : "Current user"} src={data && data.getNotifications?.profilePic} className="side-icon rounded-full"/>
      </button>
      <Link id="profile-link" to={data ? `/${data.getNotifications.handle.slice(1)}`: '/'} title="Profile" className="hidden sm:flex lg-side-icon">
        <UserIcon className="side-icon hidden sm:flex"/>
        <label htmlFor="profile-link" className="sidebar-text">Profile</label>
      </Link>
      <Link title="Sign out" id="sign-out-btn" className="hidden sm:flex lg-side-icon" to="/home" onClick={() => onSignOut()}>
        <LogoutIcon className="side-icon hidden sm:flex" id="more-icon"/>
        <label htmlFor="sign-out" className="sidebar-text">Sign Out</label>
      </Link>
    </div>
  )
}