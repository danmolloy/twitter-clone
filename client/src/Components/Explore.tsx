import { Error } from "./Error"

export const Explore = () => {
  return (
    <div className="w-full h-full" id="explore-component">
      <Error />
    </div>
  )
}