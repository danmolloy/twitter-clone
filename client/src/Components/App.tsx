import React, { useEffect, useState } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  Redirect
} from "react-router-dom"
import { Home } from './Home'
import { Explore } from './Explore'
import { Bookmarks } from './Bookmarks'
import { Messages } from './Messages'
import { Notifications } from './Notifications'
import { Profile } from './Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "./ComposeTweet";
import { RightBar }  from './RightBar'
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../constants'
import { Chat } from "./Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    follows {
      handle
    }
    followers {
      handle
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar currentUser={data && data.currentUser}/>}
        <div className="sm:ml-24 md:ml-60 mb-0 border-r w-full max-w-2xl sm:mr-2">
        <Switch>
          <Route path="/compose/tweet">
            <ComposeTweet currentUser={data && data.currentUser} updatePage={() => alert('Posted!')}/>
          </Route>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/bookmarks">
            <Bookmarks currentUser={data && data.currentUser} />
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
          <RightBar />
        </div>
    </Router>
  );
}

export default App;
