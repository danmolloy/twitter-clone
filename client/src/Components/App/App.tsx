import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from "react-router-dom"
import { Home } from '../Home/Home'
import { Explore } from '../Explore/Explore'
import { Messages } from '../Messages/Messages'
import { Notifications } from '../Notifications/Notifications'
import { Profile } from '../Profile/Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "../Home/ComposeTweet";
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../../constants'
import { Chat } from "../Messages/Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    chats {
      content {
        read
      }
    }
    follows {
      handle
    }
    followers {
      handle
    }
    notifications {
      id
      text
      sentFromUser
      read
      tweetId
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    console.log(error)
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar />}
        <div className="sm:ml-24 md:ml-60 mb-0 w-full max-w-2xl sm:mr-2 md:mr-0">
        <Switch>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
      </div>
    </Router>
  );
}

export default App;
