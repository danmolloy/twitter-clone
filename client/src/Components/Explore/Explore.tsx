import { Error } from "../App/Error"
import { gql, useQuery } from "@apollo/client"
import { UserExplore } from "./UserExplore";
import { ExploreUser } from "../../types";
import { Header } from "../App/Header";

export const ALL_USERS = gql`
  query Query {
    getAllUsers {
      name
      handle
      blurb
      profilePic
      followers {
        handle
      }
    }
  }
`;


export const Explore = (props: {currentUserHandle: string | undefined}) => {
  const {loading, error, data} = useQuery(ALL_USERS)

  return (
    <div className="w-full h-screen" id="explore-component">
      <Header pageTitle="Explore" blurb="A complete list of users"/>
      <div className="border-r min-h-screen">
      {data && data.getAllUsers.map((i: ExploreUser) => {
        return <UserExplore currentUserHandle={props.currentUserHandle && props.currentUserHandle} user={i} key={i.handle}/>
      })}
      </div>
    </div>
  )
}