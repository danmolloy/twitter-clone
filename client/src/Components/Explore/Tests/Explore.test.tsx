import { act } from "react-dom/test-utils";
import { unmountComponentAtNode, render } from "react-dom";
import { AUTH_TOKEN } from "../../../constants";
import { MockedProvider } from "@apollo/client/testing";
import { MemoryRouter } from "react-router";
import { ExploreMock, user } from "./ExploreMock";
import { Explore } from "../Explore";
import { UserExplore } from "../UserExplore";
import pretty from "pretty";
import App from '../../App/App'
import { fireEvent } from "@testing-library/dom";
import { getByText } from "@testing-library/dom";

let container: any = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
  global.localStorage.setItem(AUTH_TOKEN, "auth-token");
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  global.localStorage.removeItem(AUTH_TOKEN);
});



describe("Explore page", () => {
  // "follows or unfollows.." must be first test for mutation and succeeding snapshots to work correctly.
  it("follows or unfollows user on follow/unfollow btn click", async () => {
    act(() => {
      render(
        <MockedProvider addTypename={false} mocks={ExploreMock}>
          <MemoryRouter initialEntries={["/explore"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 200))
    })
    expect(container.textContent).toMatch(/ExploreA complete list of users/g);
    let btn = container.querySelector(".follow-unfollow-btn")
    expect(btn.textContent).toMatch(/^Unfollow$/)

    await act(async () => {
      fireEvent.click(container.querySelector(".unfollow-btn"));
      await new Promise(resolve => setTimeout(resolve, 100));
    }) 
    expect(btn.textContent).toMatch(/^Follow$/)
     
  })

  it("renders without error", async () => {
    act(() => {
      render(
        <MockedProvider addTypename={false} mocks={ExploreMock}>
          <MemoryRouter initialEntries={["/explore"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(container.textContent).toMatchSnapshot();
  })


  it("user link redirects to profile", async () => {
    act(() => {
      render(
        <MockedProvider addTypename={false} mocks={ExploreMock}>
          <MemoryRouter initialEntries={["/explore"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(container.textContent).toMatch(/ExploreA complete list of users/g);
    act(() => {
      fireEvent.click(container.querySelector("#explore-user-link"))
    })
    expect(pretty(container.innerHTML)).toMatch(/profile-component/g);
    
  })
});
