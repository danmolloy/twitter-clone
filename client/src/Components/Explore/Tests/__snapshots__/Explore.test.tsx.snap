// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore page renders without error 1`] = `"HomeExploreNotificationsMessagesProfileSign OutExploreA complete list of usersDan Molloy@danmolloyHi! Click edit profileUnfollowFiona Kelly@fizzlekellyhelloUnfollowEoghan@eggI am an eggUnfollowThe Big Firkin Band@thebigfirkinband17-piece jazz big bandUnfollow"`;
